idc_index/index.py

```diff
@@ -502,6 +465,364 @@ def get_viewer_URL(
+
+        Args:
+            output (str): The output of s5cmd sync --dry-run command.
+            downloadDir (str): The directory to download the files to.
+
+        Returns:
+            Path: The path to the generated synced manifest file.
+            float: Download size in MB
+        """
+        stdout_df = pd.DataFrame(stdout.splitlines(), columns=["s5cmd_output"])
+
+        # Extract the s3 url and crdc_instance_uuid from the s5cmd output
+        stdout_df["s3_url"] = stdout_df["s5cmd_output"].str.extract(
+            r"cp (s3://[^/]+/[^/]+)/.*"
+        )
+        stdout_df["s3_url"] = stdout_df["s3_url"] + "/*"
+        stdout_df["sync_crdc_instance_uuid"] = stdout_df["s3_url"].str.extract(
+            r"(?:.*?\/){3}([^\/?#]+)"
+        )
+        stdout_df = stdout_df[["s3_url", "sync_crdc_instance_uuid"]].drop_duplicates()
+        index_df_copy = self.index
+        index_df_copy["index_crdc_instance_uuid"] = index_df_copy[
+            "series_aws_url"
+        ].apply(self._extract_crdc_instance_uuid_from_series_aws_url)
+
+        # Perform left join on client.index using crdc_instance_uuid
+        merged_df = stdout_df.merge(
+            index_df_copy,
+            left_on="sync_crdc_instance_uuid",
+            right_on="index_crdc_instance_uuid",
+            how="left",
+        )
+        sync_size = merged_df["series_size_MB"].sum()
+        sync_size_rounded = round(sync_size, 2)
+        # Write a temporary manifest file
+        with tempfile.NamedTemporaryFile(mode="w", delete=False) as synced_manifest:
+            for s3_url in merged_df["s3_url"]:
+                synced_manifest.write(f"sync {s3_url} {downloadDir}\n")
+        return Path(synced_manifest.name), sync_size_rounded
+
+    def _s5cmd_run(
+        self, endpoint_to_use, manifest_file, total_size, downloadDir, quiet=False
+    ):
+        """
+        Executes the s5cmd command to sync files from a given endpoint to a local directory.
+
+        This function first performs a dry run of the s5cmd command to check which files need to be downloaded.
+        If there are files to be downloaded, it generates a new manifest file with the files to be synced and
+        runs the s5cmd command again to download the files. The progress of the download is tracked and printed
+        to the console.
+
+        Args:
+            endpoint_to_use (str): The endpoint URL to download the files from.
+            manifest_file (str): The path to the manifest file listing the files to be downloaded.
+            total_size (float): The total size of the files to be downloaded in MB.
+            downloadDir (str): The local directory where the files will be downloaded.
+            quiet (bool, optional): If True, suppresses the stdout and stderr of the s5cmd command. Defaults to False.
+
+        Raises:
+            subprocess.CalledProcessError: If the s5cmd command fails.
+
+        Returns:
+            None
+        """
+
+        dry_run_cmd = [
```

Comment:
And yes, I've just tested with a larger-size manifest, and the delay introduced by this s5cmd dry-run is absolutely unacceptable!

I understand we will not be able to know more precisely the size that needs to be downloaded, but IMHO it is most definitely not worth it to do dry run just to get the accurate size and suffer this enormous delay.

Reply:
Please see the svg above for why s5cmd dry run is being used.

without s5cmd sync dryrun, progress bar can sometimes be misleading. For example, when the same cohort is downloaded twice to the same folder, progress bar remains at 0%. I was just trying to implement as perfectly as possible with the tools at hand, while letting the user know when the progress bar can not be accurate.

 > dry run just to get the accurate size and suffer this enormous delay.

rsync also has a cold start right when working large data? So maybe it is normal. 

Your call on progress bar being inaccurate is okay when same cohort is downloaded more than once to the same folder vs the PR implementation.

Reply:
I do not think this enormous delay is justified to make the progress accurate. 

Also, I now looked at the documentation for `s5cmd --dry-run` https://github.com/peak/s5cmd?tab=readme-ov-file#dry-run, and it says that it "will output what operations will be performed without actually carrying out those operations". Now I am confused how it can help at all. It will just repeat the commands we already know from the manifest.

Reply:
> It will just repeat the commands we already know from the manifest.
Not true when someone interrupts the download. Once interrupted, we do not know what is already downloaded and what needs to be downloaded. Same with the case, when the download is tried twice for the same cohort to the same folder.

Example 1: If a manifest is 2 GB, and the data is already downloaded. If we run the same command again, the base line now is 2 GB, and the progress bar tries to expect 4 GB in the folder. But because the data is already there, progress bar will be at 0%

Example 2: If a manifest is 2 GB, and **SOME* of the data in the manifest is already downloaded in the destination folder. For example 1 GB. If we run the same command again, the baseline for the progress bar now is  1 GB, and the progress bar tries to expect a total of 3 GB in the folder. But because the data is already there, the progressbar will be stopped at 50%.
In these two examples, we can not provide any message to the user, except for one, `destination folder is not empty, progress bar may not always be accurate.`

With s5cmd sync dryrun we will know what needs to be actually downloaded. i.e if some data is already present, s5cmd sync dryrun output will not contain the instances that are already in the destination. We can parse the stdout from dryrun command, and calculate the sync size and download just those.

I'll create another branch without using sync dryrun I'll then create a colab notebook with all the cases where progress bar can be misleading.

Reply:
> With s5cmd sync dryrun we will know what needs to be actually downloaded.

If dry run just prints out the commands it would execute, it will not help understanding what has already been downloaded. There file may be zero size, or incomplete, or may have content different from remote. Communication with the server will be needed to confirm whether something was completed or not, and for that you need more than dry run. I have not run any experiments, but that's my understanding based on the documentation. So I do not think it solves the problem.

> I'll create another branch without using sync dryrun I'll then create a colab notebook with all the cases where progress bar can be misleading.

I really do not see why this should be a priority at all.

You have to weigh perfection vs impact.

If the download is partially complete, your estimate will be larger than the real size to be downloaded. So the result will be that user will in reality wait less time than projected. What is the negative impact with that from the user perspective? I see no problem. In your first example, it will jump from 0 to done very quickly, in the second example user will wait half the time. Zero negative impact IMHO.

At the same time, if we seek perfection, every user will suffer from the increased delay for every download. This is clearly a significant problem for the user. So it does not matter how dry run works. Bottom line - it increases download time by a lot for even a moderate size manifest of 65K, and it is not worth the benefit that brings (even if dry run can give us that estimate).

Reply:
> If dry run just prints out the commands it would execute, it will not help understanding what has already been downloaded. There file may be zero size, or incomplete, or may have content different from remote. Communication with the server will be needed to confirm whether something was completed or not, and for that you need more than dry run. I have not run any experiments, but that's my understanding based on the documentation. So I do not think it solves the problem.

Unfortunately, this is not true. You are stressing only on the dry-run and not the sync. Please see the documentation for how s5cmd handles sync. Sync checks what is already in the download folder by two parameters size and time of the file modified (i.e. when IDC ingested them to AWS).
https://github.com/peak/s5cmd?tab=readme-ov-file#sync

 > In your first example, it will jump from 0 to done very quickly
 Progress bar remains at 0% not jump to 100%.
 
> At the same time, if we seek perfection, every user will suffer from the increased delay for every download. This is clearly a significant problem for the user. So it does not matter how dry run works. Bottom line - it increases download time by a lot for even a moderate size manifest of 65K, and it is not worth the benefit that brings (even if dry run can give us that estimate).

In my opinion, we should implement the best way we possibly can. For me, misleading progress bar is worse than the cold start. But if you are fine with that, I'll be too.

I can think of one more solution:
1. Offer progress bar as an optional feature. Because for cases like SlicerIDCBrowser, progress bar may not be helpful at all.
2.  If the destination folder is empty, I'll agree with you that there is no need to sync dry-run.
3. Only when the user chooses progressbar=True, and the destination folder is not empty, we can do s5cmd sync dryrun.

Reply:
> For me, misleading progress bar is worse than the cold start. 

Why is that?

I had another thought about how we can do it. Why do we need to check with the server at all. We can loop just over the local destination files in the manifest and calculate their size locally. It will be most definitely faster than the current dry run! I am not sure why we didn't consider this earlier.

Reply:
Because I know of at least three cases where progress bar will be misleading. Who knows how many other possibilities there are. I discussed two cases discussed above. Third is when there are some and not all instances are removed for a series. Because we have only series level info, we can't calculate the partial size to be downloaded. This is an inherent problem with our data model and nothing much we can do about it.


I don't understand what you mean about looping over the dicom files in the destination folder. 

As you know, manifest does not have instances info, only series folder. How can we know the size (when the folder contains for example unrelated dicom files to the manifest in the destination), what is downloaded and what is not?   

Please help me understand.

Reply:
> Because I know of at least three cases where progress bar will be misleading.

I do understand that. But I meant to ask why do you consider accurate progress bar to be more important than cold start? 

> As you know, manifest does not have instances info, only series folder. 

The decision to flatten the content of the series dirs is not rooted in user needs. There is no harm (quite the opposite) to keep series-level folders at the destination. If we have the list of series folders to track, everything becomes much easier.

But let's wrap up this PR and address this separately.

Reply:
> I do understand that. But I meant to ask why do you consider accurate progress bar to be more important than cold start?

because unless we are benchmarking or using for using it for viewer applications such as SlicerIDCBrowser, speed may not be a priority. Accuracy may be. If I'm downloading 10 TB and the download is interrupted for whatever reason and if I'm counting on the progress bar to track progress, I would appreciate the accuracy. Again, we don't have to force the progress bar at all times. We could implement it as a feature. For those who value speed over an accurate progress bar, they absolutely can disable the feature as I mentioned above.

> The decision to flatten the content of the series dirs is not rooted in user needs. There is no harm (quite the opposite) to keep series-level folders at the destination. If we have the list of series folders to track, everything becomes much easier.

This is unfortunately not foolproof. Once again if some data is already present in the destination folder even with series folder implementation, yes, we may know the size of the destination series folder is different. But we will need to think about how to address the inconsistencies. We will essentially be implementing sync logic when it is already available in s5cmd.

> But let's wrap up this PR and address this separately.

Could I know if you have any objections to this?

>1. Offer progress bar as an optional feature. Because for cases like SlicerIDCBrowser, progress bar may not be helpful at all.
> 2. If the destination folder is empty, I'll agree with you that there is no need to sync dry-run.
> 3. Only when the user chooses progressbar=True, and the destination folder is not empty, we can do s5cmd sync dryrun.

Reply:
> because unless we are benchmarking or using for viewer applications such as SlicerIDCBrowser, speed may not be a priority.

In my view, speed is always the priority. One of the modes of operation for SlicerIDCBrowser is interactive downloader, so it will be relevant there as well. Based on the arguments I have seen, I would like to go ahead with prioritizing speed over accuracy of progress. If user concerns emerge over time, I am ready to revisit. If you like, we can bring this up at the weekly meeting tomorrow.

> Offer progress bar as an optional feature. Because for cases like SlicerIDCBrowser, progress bar may not be helpful at all.

This is actually a good point we did not consider earlier. Progress reporting will most definitely be very important for SlicerIDCBrowser. We need a function that will be able to help us update the progress bar in Slicer. Can you look into how progress bar is implemented in Slicer typically, and see what we might need to do to support it with this function?

> Only when the user chooses progressbar=True, and the destination folder is not empty, we can do s5cmd sync dryrun.

I think both progress bar and dryrun should be optional. Whenever we get to providing a command line download tool, progress should be enabled by default, and dryrun should be disabled by default.

Reply:
>In my view, speed is always the priority. One of the modes of operation for SlicerIDCBrowser is interactive downloader, so it will be relevant there as well. 
This is actually a good point we did not consider earlier. Progress reporting will most definitely be very important for SlicerIDCBrowser. We need a function that will be able to help us update the progress bar in Slicer. Can you look into how progress bar is implemented in Slicer typically, and see what we might need to do to support it with this function?


I'm totally aware of need to bring progress bar to SlicerIDCBrowser. The reason why I kept saying the progress bar implemented here will be not useful in SlicerIDCBrowser is because, progress bar is displayed with qt there and idc-index's progress bar will be of no use. We will need to implement it again there.

https://github.com/PerkLab/SlicerSandbox/blob/master/LoadRemoteFile/LoadRemoteFile.py#L50

> I think both progress bar and dryrun should be optional. Whenever we get to providing a command line download tool, progress should be enabled by default, and dryrun should be disabled by default.

I do not understand what you finally want here. Please correct me if my understanding is wrong. For now, everyone is opted into progress bar and no option to select whether to enable or disable. Second do not use sync dryrun at all? Is that correct?

Reply:
Ok. Sorry for the confusion, let's not increase the complexity. Forget the other points. Let's do the following:

1. Make both progress reporting and dry run configurable via function parameters.
2. Progress reporting is enabled by default
3. Dry run is disabled by default

Reply:
All requested constraints are now met.

I am not a fan and find it unethical to provide an option to skip validating manifests. As s5cmd does not interrupt download even with invalid URLs. There is no meaningful savings of any kind by skipping. But I'll leave it as it is.

![image](https://github.com/ImagingDataCommons/idc-index/assets/115020590/64e8d53e-9cd6-4ab8-8342-377d88b17ebb)

Resolving this.